youtube-dl: add man page, completion data, python subpackages

fbrosson opened the pull request:
Changes:

* Bump to 2018.02.11.
* Add man page and bash/fish/zsh completion data.
* Fix/update `LICENSE` & `COPYRIGHT`.
* Drop `cmd:ffmpeg` from `REQUIRES` as ffmpeg is not required but just recommended. youtube-dl only needs ffmpeg if the user downloads DASH video+audio contents and wants these to be combined together.
* Add _python and _python3 subpackages.

Contents:

* `youtube_dl` ships the `youtube-dl` monolithic python2.7 script, as well as the man page and the bash/fish/zsh completion data.
* `youtube_dl_python` ships the `youtube-dl-python` python2.7 script.
* `youtube_dl_python3` ships the `youtube-dl-python3` python3.6 script.

Remarks:    

* The base package, youtube_dl, does not depend on any of the python subpackages.
* The youtube_dl_python and youtube_dl_python3 packages do not depend on the base package.
* All packages may be installed together.
* Each package works independently of the other two.
* youtube-dl-python3 does not work on x86_gcc2 because python3 is not available on x86_gcc2 primary arch and something is preventing youtube-dl-python3 from working correctly with a secondary arch build of python3. So youtube_dl_python3 is not built on x86_gcc2.

alaviss commented on net-misc/youtube_dl/youtube_dl-2018.02.08.recipe:
```diff
@@ -26,20 +26,18 @@ REQUIRES="
-	sed -i 's/\/usr//' youtube-dl-$portVersion.py
-	mv youtube-dl-$portVersion.py youtube-dl
-	chmod a+x youtube-dl
-}
```
The port is a bit more complicated. Looks like this port is a perfectly valid python module that uses setuptools. This basically means we would have to use this ugly "pattern"
https://github.com/haikuports/haikuports/blob/99b5ce0a6ffec38d8db4304ec4fc2f58108a9253/dev-python/docutils/docutils-0.14.recipe#L31

Why should we further complicate this port? So that youtube-dl could be [embedded][0] to a python program.

[0]: https://github.com/rg3/youtube-dl/tree/2018.02.08#embedding-youtube-dl

fbrosson replied:
That's a good idea! Thanks for the suggestion!
(I'll try to implement this before tomorow.)

pulkomandy commented:
> > @@ -9,9 +9,9 @@ HOMEPAGE="http://rg3.github.io/youtube-dl/"
>  COPYRIGHT="2006-2018 Ricardo Garcia Gonzalez"
> 
> There's no copyright for a project licensed to the public domain. It appears that there's two popular value for this case: "None" or "Public Domain"
> 
> > @@ -9,9 +9,9 @@ HOMEPAGE="http://rg3.github.io/youtube-dl/"
>  COPYRIGHT="2006-2018 Ricardo Garcia Gonzalez"
>  LICENSE="Public Domain"
> 
> They don't release the project to the public domain using CC0 (aka "Public Domain" in `/system/data/licenses`) but instead they use ["Unlicense"][0]
> 
> [0]: https://github.com/rg3/youtube-dl/blob/2018.02.08/LICENSE

If it uses CC0 or Unlicense, then it is not exactly public domain. And
the only difference is precisely that there still is a copyright. The
effect of this is minor and only relevant in some legislations.

In France, things can be in the public domain only 50 years after the
author's death. There is no way to shorten this even if the author
wishes to. CC0 and Unlicense are as close as it gets.

As a result, even if something is in the public domain, it means there
was, at some point, a copyright. And we should mention it, because for
legal purposes it is the expiration of that copyright that allows us to
say that the software is in the public domain.

fbrosson commented on net-misc/youtube_dl/youtube_dl-2018.02.08.recipe:
```diff
@@ -9,9 +9,9 @@ HOMEPAGE="http://rg3.github.io/youtube-dl/"
-SOURCE_URI="https://github.com/rg3/youtube-dl/releases/download/$portVersion/youtube-dl#noarchive"
-CHECKSUM_SHA256="157be71476a97dae99a18df5452472e1786f39b3fe2f968dfa0f81c144d81734"
-SOURCE_FILENAME="youtube-dl-$portVersion.py"
+SOURCE_URI="https://github.com/rg3/youtube-dl/releases/download/$portVersion/youtube-dl-$portVersion.tar.gz"
```
Thanks for the info. I had not noticed the full tarball was also available from the upstream website.

That said, I think it is nicer to have an https URI instead of an http URI, even if the latter is shorter than the former.

alaviss replied:
@fbrosson Thanks for the pointer. The upstream site also have https :stuck_out_tongue:

fbrosson replied:
Cool! I'll switch `HOMEPAGE` to https!

BTW, your `SOURCE_URI` sends a redirect (`HTTP/1.1 302 Found`) to GitHub's URI.
So we could use any of them.
But since we are getting a "302" (and not a "301"), I'll switch to **_your_** `SOURCE_URI` :)
(Your URI is better, as you already know, because it can save us an update in case upstream ever chooses to switch to another hosting service.)

fbrosson commented:
OK, thanks @alaviss and @pulkomandy.

1) Shall I change `LICENSE` to `Unlicense` (and place a copy of that file to the `licenses` directory) ?

2) Shall I keep `COPYRIGHT` as it is now, or should I replace `Ricardo Garcia Gonzalez` by `Ricardo Garcia Gonzalez and other Youtube-dl authors` (or just `Youtube-dl authors`) and include the `AUTHORS` file in `$docDir` ?

alaviss commented:
@pulkomandy So who should hold the copyright if that's the case? No one has claimed the copyright ever since the project [started][0]

[0]: https://github.com/rg3/youtube-dl/blob/4fa74b5252a23c2890ddee52b8ee5811b5bb2987/youtube-dl#L4

pulkomandy commented:
On Sat, Feb 10, 2018 at 05:09:18PM +0000, alaviss wrote:
> @pulkomandy So who should hold the copyright if that's the case? No one has claimed the copyright ever since the project [started][0]

Something generic like "youtube-dl contributors" then?

fbrosson commented:
I've made the changes suggested by @alaviss to allow youtube-dl's functionality to be available to python scripts. Here are the contents of the 3 packages created by the recipe:

package | contents
--- | ---
youtube_dl | `youtube-dl` monolithic python2 script,<br/>man page<br />bash/fish/zsh completion data.
youtube_dl_python | `youtube-dl-python` python2 script using the py/pyc files.
youtube_dl_python3 | `youtube-dl-python3` python3 script using the py/pyc files.

Remarks:
* The base package, `youtube_dl`, does not depend on any of the python subpackages.
* The `youtube_dl_python` and  `youtube_dl_python3` packages do not depend on the base package.
* All packages may be installed together.
* Each package works independently of the other two.
* `youtube_dl_python3` is not built on x86_gcc2 because `youtube-dl-python3` does not work with python3_x86.

**Questions:**
1) Should I build `youtube_dl_python3` even on x86_gcc2?
2) May I merge this PR?

fbrosson commented:
I just fixed a typo in a comment in the recipe.
While at it, I also added a detailed description for one of the commits.
(But I'll use a slightly different commit summary, see below.)

Any suggestions (on both the code and commit summary) are welcome.
Here's what I plan to use as commit summary, for the squash and merge:
```
youtube-dl: add man page, completion data, python subpackages

Changes:

* Bump to 2018.02.11.
* Add man page and bash/fish/zsh completion data.
* Fix/update LICENSE & COPYRIGHT.
* Add _python and _python3 subpackages.

Contents:

* youtube_dl ships the main youtube-dl monolithic python2 script, as
  well as the man page and the bash/fish/zsh completion data.
* youtube_dl_python ships the youtube-dl-python python2 script.
* youtube_dl_python3 ships the youtube-dl-python3 python3 script.

Remarks:    

* The base package, youtube_dl, does not depend on any of the python
  subpackages.
* The youtube_dl_python and youtube_dl_python3 packages do not depend
  on the base package.
* All packages may be installed together.
* Each package works independently of the other two.
* youtube-dl-python3 does not work on x86_gcc2 because python3 is not
  available on x86_gcc2 primary arch and something is preventing
  youtube-dl-python3 from working correctly with a secondary arch
  build of python3. So youtube_dl_python3 is not built on x86_gcc2.
```

alaviss commented:
Sorry for not being around. It's the holidays at my place.

Here are a few suggestions:
- Don't use the Makefile. Python setuptools has already done everything needed.
- Pick a python version to use as default. Probably you could just remove $binDir/python-dl for the other version, as both should act the same.
- You should check the fish recipe to see where completions were put

On February 12, 2018 8:48:25 PM GMT+07:00, fbrosson <notifications@github.com> wrote:
>@fbrosson requested your review on: haikuports/haikuports#2202
>youtube-dl: add man page, completion data, python subpackages.
>
>-- 
>You are receiving this because your review was requested.
>Reply to this email directly or view it on GitHub:
>https://github.com/haikuports/haikuports/pull/2202#event-1469474366

fbrosson commented:
Thanks @alaviss for your suggestions!

(I had not dared to drop the monolithic python script, but, after all, this seems to be a good idea.)

So, here's what I plan to do (if everyone is OK with dropping the monolithic python script):
* Drop the `youtube-dl` from the `youtube_dl` base package.
* Move/merge the contents of the `youtube_dl_python` subpackage to the `youtube_dl` base package, renaming `youtube-dl-python` to `youtube-dl`.
* Drop the `youtube-dl-python3` script from the `youtube_dl_python3` subpackage.

**Questions:**
1) Should I build `youtube_dl_python3` on x86_gcc2 ? (I guess not.)
2) Should I rename `youtube_dl_python3` to `youtube_dl3` ? (I guess not.)
3) When not on x86_gcc2, should I merge into `youtube_dl` the contents of `youtube_dl_python3` (and add a dependency on python3.6) ? (I guess not.)
4) Is everyone OK with the suggested changes? I'm asking this because these changes will break the small similarity we have between the current `youtube_dl_python` and `youtube_dl_python3` subpackages.

Thanks!

korli commented:
I'd like to keep one standalone youtube_dl package, I don't see the need for the subpackages.

fbrosson commented:
Thanks @korli for your comment.

I must say that I was not very happy with an asymetric split.
(BTW, I'm glad i had not started this change yet.)

I do agree that, for this recipe, having only one package is better than having subpackages.

I see 3 possibles scenarios:
1) We drop the `youtube_dl_python` and `youtube_dl_python3` subpackages, and only keep the monolithic python script in the (base) package.
2) We merge `youtube_dl_python` into `youtube_dl` and drop `youtube_dl_python3`.
3) Same as 2 on x86_gcc2, but, on all other architectures, we merge both `youtube_dl_python` and `youtube_dl_python3` into `youtube_dl`.

For scenarios 2 and 3, I guess it would be nice to drop the monolithic script and use the tiny entry-point python2.7 script, calling it `youtube-dl`, of course.

I think scenarios 2 or 3 would be nicer than scenario 1, but I do agree that scenario 3 might not be acceptable (as it would force a double dependency on both python2.7 and python3.6.)
I'm also OK with scenario 1 (which was what I had initially implemented.)
BTW, I don't have any preference, so I'm open to all scenarios.

fbrosson commented:
> I'd like to keep one standalone youtube_dl package

scenario 1 = standalone `youtube-dl` **_script_** (with a dependency on python2.7).
scenario 2 = standalone `youtube_dl` **_package_** (with a dependency on python2.7).
scenario 3 = standalone `youtube_dl` **_package_** (with a dependency on both python2.7 and python3.6).

@korli, would you be OK with scenario 2 (or maybe 3)?
Or do you prefer to have a standalone `youtube-dl` **_script_** ?

Thanks!

korli commented:
I suppose scenario 2 is OK, provided a script is still in /bin. Thanks.

korli commented:
fbrosson, is there any existing package which depends on youtube-dl python module? I bet it's a no (nothing on AUR for instance).

fbrosson commented:
Nope.

BTW, there is only one port which needs youtube-dl: [haiku-apps/ubertuber](https://github.com/haikuports/haikuports/tree/master/haiku-apps/ubertuber), but it only depends on `cmd:youtube_dl`, so there is no need for anything that would be provided by any of the subpackages.

fbrosson commented:
If the 4 subpackages in scenario 4 seem too much, would it be OK to limit that to 2 optional subpackages?

This is exactly what we currently have in this PR, as I wrote in [this comment](#issuecomment-364914375).
Let's call this scenario 5, (although it's there since yesterday :)
And let's call scenario 0 what we would get if we only kept the base package of scenario 4 (which happens to be the same as the base package of scenario 5.)

So we can choose between these scenarios:

\# | Comments | Work
-- | -- | --
0 | A single package, with the monolithic script, plus the man page and the completion data. | Very easy.
2 | Same as scenario 0 but with a tiny script instead of the monolithic script, and with all py&pyc files for python2.7. | Big changes
4 | Base package identical to that of scenario 0, but with 4 optional subpackages. | Easy changes.
5 | Base package identical to that of scenario 0, but with 2 optional subpackages. | **_Ready_.**
6 | Same as scenario 5, but without `youtube-dl-python` and `youtube-dl-python3` | Very easy.

fbrosson commented:
@korli, although [you wrote](#issuecomment-365221971) that you were OK with scenario 2, I think what you would prefer is scenario 0, i.e. a single package with the monolithic script, with the man page and completion data. Am I right?
Well, since there does not seem to be a consensus on anything other than scenario 0, I'm OK with it.

@alaviss & @pulkomandy (and others too), would you also be OK with this scenario 0?
(It's really easy to switch to that scenario. All I have to do is drop the fourth and fifth commits in this PR.)

So please let me know if you are all OK with scenario 0, or if you would prefer to wait until we reach a consensus on another scenario.
Thanks!

Shall we do a small votation with thumbs up or down, to each of these scenarios?
(You can put 👍 or 👎 for several scenarios.)

fbrosson commented:
**Scenario 0:** A single package, with the monolithic script, plus the man page and the completion data.

fbrosson commented:
**Scenario 2:** Same as scenario 0 but with a tiny script instead of the monolithic script, and with all py&pyc files for python2.7.

fbrosson commented:
**Scenario 4:** Base package identical to that of scenario 0, but with 4 optional subpackages.

fbrosson commented:
**Scenario 5:** Base package identical to that of scenario 0, but with 2 optional subpackages.

fbrosson commented:
**Scenario 6:** Same as scenario 5, but without youtube-dl-python and youtube-dl-python3.

fbrosson commented:
Let's add the "WIP-no-merge" label to this PR, until we reach a nice consensus.

(Please continue "voting" to choose those scenarios that you prefer.)

BTW, in case we go for scenario 5, we have a text in [this comment](#issuecomment-364922589) which can serve as commit summary. But we need to copy paste that text when we do the squash & merge.
(Well, I can squash all commits together and "preset' the correct commit message. But it's better to wait until we know what scenario we choose.)

So what I suggest is that I merge this PR myself once we have a nice consensus.

fbrosson commented:
@korli & @pulkomandy, please do not hesitate to vote for scenario 2, like @alaviss did, or for scenario 4 like @waddlesplash did, or for any other scenario, of course.
BTW, the changes for scenario 4 won't be that big, after all.
(I voted for scenario 4 because it's the most flexible solution, imho.)

Reminder: all users reading this comment are welcome to vote for any scenario, and also change any previous vote (by clicking again on the 👍and then choosing another scenario.)

pulkomandy commented:
On Thu, Feb 15, 2018 at 06:45:44PM +0000, fbrosson wrote:
> @korli & @pulkomandy, please do not hesitate to vote for scenario 2, like @alaviss did, or for scenario 4 like @waddlesplash did, or for any other scenario, of course.
> BTW, the changes for scenario 4 won't be that big, after all.
> (I voted for scenario 4 because it's the most flexible solution, imho.)

I did not vote because I think this does not make any difference, so it
is not worth me reviewing things and deciding on a solution.

fbrosson commented:
No problem. Thanks for your comment :)
Let's go for scenario 4 then, since it already has 3 👍, while scenario 2 only has 1 👍.
